LeaderboardBackend/Services/Impl/AccountConfirmationService.cs

```diff
@@ -48,18 +49,18 @@ public async Task<CreateConfirmationResult> CreateConfirmationAndSendEmail(User
 
         _applicationContext.AccountConfirmations.Add(newConfirmation);
 
-        await _applicationContext.SaveChangesAsync();
-
         try
         {
             await _emailSender.EnqueueEmailAsync(
                 user.Email,
                 "Confirm Your Account",
                 GenerateAccountConfirmationEmailBody(user, newConfirmation)
             );
+            await _applicationContext.SaveChangesAsync();
         }
         catch
         {
+            _applicationContext.AccountConfirmations.Entry(newConfirmation).State = EntityState.Detached;
             // TODO: Log/otherwise handle the fact that the email failed to be queued - zysim
             return new EmailFailed();
         }
```

Comment:
You're making the assumption here that the exception was thrown when attempting to send the email and not while saving to the database. If you want to ensure that this is the case, catch a specific exception. In fact, I don't think this is going to work at all since `EnqueueEmailAsync` doesn't seem to throw errors.

Reply:
Here's an excerpt from `EnqueueEmailAsync`. As you can see, any exceptions generated by `SendEmail` are swallowed.

```
try
{
    await SendEmail(message);
}
finally
{
    _semaphore.Release();
}
```

Reply:
Or maybe not? I guess they're automatically propagated if there's no catch block?

Reply:
The previous code was better imo.
As things are right now, it's better to save the confirmation before sending the email. If you send the email and then the DB request fails, the user is getting a broken link in their email.

(once email queueing is implemented, both actions can be in the same transaction and fail together)

Reply:
I think in that case it's better to just create the confirmation object outside of the try catch then but do it BEFORE. We really don't expect it to fail because if it does then we have bigger issues.

Reply:
We really don't care about the case when the email fails but the object gets created. You could just not even handle it.

Reply:
I figured that we wouldn't want to create a confirmation in the table if the email fails to send (and of course if saving to DB fails, then it wouldn't be added at all). But yeah; forgot about the fact that this change means that an email can be sent even if the DB transaction fails. I'll revert the changes then

Reply:
If it fails to save changes to the DB, I believe it will through an exception, which will result in the whole thread essentially erroring out and terminating before it even gets to sending the email.